### PR TITLE
SNOW-641125: support put_stream from sp-python-connector

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-641125

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [x] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Enabled support for `options` for `_upload_stream` in stored proc python connector in [this](https://github.com/snowflakedb/Stored-Proc-Python-Connector/pull/61) and [this](https://github.com/snowflakedb/Stored-Proc-Python-Connector/pull/63) PR. Now exposing the support to python client.

### CODE REVIEW

File: src/snowflake/snowpark/file_operation.py [hunk 1]
```diff
@@ -239,22 +239,25 @@ def put_stream(
...
+
         if is_in_stored_procedure():
-            raise NotImplementedError(
-                "Stream uploads for stored procedure are not supported yet"
-            )
+            options = {
+                "parallel": parallel,
+                "source_compression": source_compression,
+                "auto_compress": auto_compress,
+                "overwrite": overwrite,
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/file_operation.py:253]: Do we have test code to cover all these options?

File: src/snowflake/snowpark/file_operation.py [hunk 2]
```diff
@@ -239,22 +239,25 @@ def put_stream(
...
+            }
+            cursor = self._session._conn._cursor
+            cursor._upload_stream(input_stream, stage_location, options)
+            result_data = cursor.fetchall()
+            return PutResult(*result_data[0])
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/file_operation.py]: we should. updated

File: src/snowflake/snowpark/file_operation.py [hunk 3]
```diff
@@ -239,22 +239,27 @@ def put_stream(
         Returns:
             An object of :class:`PutResult` which represents the results of an uploaded file.
         """
+        stage_location = stage_location.strip()
+        if not stage_location:
+            raise ValueError("stage_location cannot be empty")
+        if stage_location[-1] == "/":
+            raise ValueError("stage_location should end with target filename")
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/file_operation.py:246]: How about adding what's a correct value for stage_location after telling them what's wrong?
For instance, "stage_location cannot be empty. It must be a full stage path with prefix and file name..., like <the correct format>"

File: src/snowflake/snowpark/file_operation.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/file_operation.py:257]: pretty much. `overwrite` and `auto_compress` are the main ones. I also added a test to see if auto detection for source compression works with gzip.

File: src/snowflake/snowpark/file_operation.py [hunk 4]
```diff
@@ -239,22 +239,31 @@ def put_stream(
...
 
-            stage_with_prefix, dest_filename = parse_stage_file_location(stage_location)
+        if is_in_stored_procedure():
+            options = {
+                "parallel": parallel,
+                "source_compression": source_compression,
+                "auto_compress": auto_compress,
+                "overwrite": overwrite,
+            }
+            cursor = self._session._conn._cursor
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/file_operation.py:259]: Do we also need the try-except similar to `put` so that the exception is wrapped?

**Review by sfc-gh-sfan — APPROVED** [review]
One comment. Otherwise LGTM

File: src/snowflake/snowpark/file_operation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/file_operation.py]: Can we also update for non-stored proc case in this function? Sorry I didn't notice it earlier in previous PR.

[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/file_operation.py]: I'm not sure if we can. This is the raw result if get after upload stream
`{'data': [('something.txt', 'something.txt.gz', 9, 0, 'NONE', 'GZIP', 'SKIPPED', '')], 'sfqid': '01a6a08a-0505-e288-0000-90030b2b3696'}`

**Review by sfc-gh-yixie — APPROVED** [review]

File: src/snowflake/snowpark/file_operation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/file_operation.py]: hmm here you can do the same:
```
result_data = result_of_upload_stream["data"]
result_meta = cursor.description
put_result = result_set_to_rows(result_data, result_meta)[0]
return PutResult(**put_result.asDict())
```
(then you can combine common code for SP and non-SP case)

[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/file_operation.py]: the `upload_stream` on non-SP side does not use cursor directly. cursor is wrapper somewhere underneath and we use `run_query` instead

https://github.com/snowflakedb/snowpark-python/blob/main/src/snowflake/snowpark/_internal/server_connection.py#L311

[the remark this answers is not in the document]
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/file_operation.py]: I see, then let's just keep the original implementation. Thanks!

**Review by sfc-gh-jdu — APPROVED** [review]

File: src/snowflake/snowpark/file_operation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/file_operation.py]: I was wrong. It is possible to access cursor. you code suggestion works. I'll add it to the PR

[the remark this answers is not in the document]
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/file_operation.py]: Just tossing ideas that is beyond this PR's scope: maybe we should consider moving the `if-else` block for these methods to an internal util file such that it is cleaner in the public file.

**Pull request merged**